### PR TITLE
ddl,statistics: handle the drop database event in the PQ

### PR DESCRIPTION
**Pull request by Rustin170506**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/57227

Problem Summary:

### What changed and how does it work?

We forget to handle the drop schema event in the PQ. So in this PR, we used the DDL notifier to deliver the drop schema event to the PQ. In the PQ, we try to delete the table jobs for the tables that come from this schema.

To avoid a transition that is too large, I created some mini-structs to include only the essential information for the schema event.

I also split the job into smaller jobs so that the subscriber can avoid performing a bulk update.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/notifier/events.go [hunk 1]
```diff
@@ -426,9 +426,28 @@ func NewFlashbackClusterEvent() *SchemaChangeEvent {
...
+func (s *SchemaChangeEvent) GetDropSchemaInfo() (dbInfo *model.DBInfo, tables []*model.TableInfo) {
+	intest.Assert(s.inner.Tp == model.ActionDropSchema)
+	return s.inner.DBInfo, s.inner.Tables
+}
+
 // jsonSchemaChangeEvent is used by SchemaChangeEvent when needed to (un)marshal data,
 // we want to hide the details to subscribers, so SchemaChangeEvent contain this struct.
 type jsonSchemaChangeEvent struct {
+	DBInfo          *model.DBInfo        `json:"db_info,omitempty"`
+	Tables          []*model.TableInfo   `json:"tables,omitempty"`
```
Comment (lance6716) [inline pkg/ddl/notifier/events.go:450]: this event can be very large. Can we only save needed information?

**Review by lance6716 — APPROVED** [review]
lgtm for DDL part

File: pkg/ddl/notifier/events.go (hunk 1 above)
Comment (Rustin170506) [reply pkg/ddl/notifier/events.go:450]: The partitioned information is needed. Maybe we should consider only store ids?

**Review by fzzf678 — APPROVED** [review]

**Comment by Rustin170506** [comment]
/hold

The txn is too big. We should consider other solutions.

File: pkg/ddl/notifier/events.go (hunk 1 above)
Comment (Rustin170506) [reply pkg/ddl/notifier/events.go]: Updated.

**Review by Rustin170506 — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Review by time-and-fate — APPROVED** [review]

File: pkg/ddl/notifier/events.go [hunk 2]
```diff
@@ -426,15 +427,77 @@ func NewFlashbackClusterEvent() *SchemaChangeEvent {
...
+}
+
+// GetDropSchemaInfo returns the database info and tables of the SchemaChangeEvent whose type is ActionDropSchema.
+func (s *SchemaChangeEvent) GetDropSchemaInfo() (miniDBInfo *MiniDBInfoForSchemaEvent) {
+	intest.Assert(s.inner.Tp == model.ActionDropSchema)
+	return s.inner.MiniDBInfo
+}
+
+// MiniDBInfoForSchemaEvent is a mini version of DBInfo for DropSchemaEvent only.
+type MiniDBInfoForSchemaEvent struct {
```
Comment (lance6716) [inline pkg/ddl/notifier/events.go:466]: We have `TableNameInfo` in "model" package, I also prefer define this struct there.

File: pkg/ddl/schema.go [hunk 3]
```diff
@@ -218,6 +219,23 @@ func onDropSchema(jobCtx *jobContext, job *model.Job) (ver int64, _ error) {
 			break
 		}
 
+		// Split tables into multiple jobs to avoid too big transaction.
```
Comment (lance6716) [inline pkg/ddl/schema.go:222]: Although you split to multiple `asyncNotifyEvent`, DDL still uses the same transaction to commit. I think you means "...avoid too big records"

File: pkg/ddl/notifier/events.go [hunk 4]
```diff
@@ -426,15 +427,77 @@ func NewFlashbackClusterEvent() *SchemaChangeEvent {
...
+func NewDropSchemaEvent(dbInfo *model.DBInfo, tables []*model.TableInfo) *SchemaChangeEvent {
+	miniTables := make([]*MiniTableInfoForSchemaEvent, len(tables))
+	for i, table := range tables {
+		miniTables[i] = &MiniTableInfoForSchemaEvent{
+			ID:   table.ID,
+			Name: table.Name,
+		}
+		if table.Partition != nil {
+			for _, part := range table.Partition.Definitions {
+				miniTables[i].Partitions = append(miniTables[i].Partitions, &MiniPartitionInfoForSchemaEvent{
```
Comment (lance6716) [inline pkg/ddl/notifier/events.go:440]: you have prepare the slice capacity in line 432 but not here 🤔 I think a table may have many partitions so it's also good to do it.

**Review by lance6716 — COMMENTED** [review]
rest lgtm for DDL part

File: pkg/ddl/notifier/events.go (hunk 4 above)
Comment (Rustin170506) [reply pkg/ddl/notifier/events.go]: Updated.

File: pkg/ddl/notifier/events.go (hunk 2 above)
Comment (Rustin170506) [reply pkg/ddl/notifier/events.go:468]: This is schema only event. So I don't think we should put it there.

File: pkg/ddl/schema.go (hunk 3 above)
Comment (Rustin170506) [reply pkg/ddl/schema.go]: Upatded.

**Review by Rustin170506 — COMMENTED** [review]
🔢 Self-check (PR reviewed by myself and ready for feedback.)

**Review by fzzf678 — APPROVED** [review]

**Pull request merged**